AppSandboxFileAccess/Classes/AppSandboxFileAccess.m

```diff
@@ -111,38 +114,80 @@ - (NSURL *)askPermissionForUrl:(NSURL *)url {
 	}
 
-	return allowedUrl;
+	return allowedURL;
 }
 
-- (void)persistPermissionPath:(NSString *)path {
-	[self persistPermissionURL:[NSURL fileURLWithPath:path]];
+- (NSData *)persistPermissionPath:(NSString *)path {
+	NSParameterAssert(path);
+	
+	return [self persistPermissionURL:[NSURL fileURLWithPath:path]];
 }
 
-- (void)persistPermissionURL:(NSURL *)url {
+- (NSData *)persistPermissionURL:(NSURL *)url {
+	NSParameterAssert(url);
+	
 	// store the sandbox permissions
 	NSData *bookmarkData = [url bookmarkDataWithOptions:NSURLBookmarkCreationWithSecurityScope includingResourceValuesForKeys:nil relativeToURL:nil error:NULL];
 	if (bookmarkData) {
 		[AppSandboxFileAccessPersist setBookmarkData:bookmarkData forURL:url];
 	}
+	return bookmarkData;
 }
 
 - (BOOL)accessFilePath:(NSString *)path withBlock:(AppSandboxFileAccessBlock)block persistPermission:(BOOL)persist {
-	return [self accessFileURL:[NSURL fileURLWithPath:path] withBlock:block persistPermission:persist];
+	// Deprecated. Use 'accessFilePath:persistPermission:withBlock:' instead.
+	return [self accessFilePath:path persistPermission:persist withBlock:block];
+}
+
+- (BOOL)accessFileURL:(NSURL *)fileURL withBlock:(AppSandboxFileAccessBlock)block persistPermission:(BOOL)persist {
+	// Deprecated. Use 'accessFileURL:persistPermission:withBlock:' instead.
+	return [self accessFileURL:fileURL persistPermission:persist withBlock:block];
+}
+
+- (BOOL)accessFilePath:(NSString *)path persistPermission:(BOOL)persist withBlock:(AppSandboxFileAccessBlock)block {
+	return [self accessFileURL:[NSURL fileURLWithPath:path] persistPermission:persist withBlock:block];
+}
+
+- (BOOL)accessFileURL:(NSURL *)fileURL persistPermission:(BOOL)persist withBlock:(AppSandboxFileAccessBlock)block {
+	NSParameterAssert(fileURL);
+	NSParameterAssert(block);
+	
+	BOOL success = [self requestAccessPermissionsForFileURL:fileURL persistPermission:persist withBlock:^(NSURL *securityScopedFileURL, NSData *bookmarkData) {
+		// execute the block with the file access permissions
+		@try {
+			[securityScopedFileURL startAccessingSecurityScopedResource];
+			block();
+		} @finally {
+			[securityScopedFileURL stopAccessingSecurityScopedResource];
+		}
+	}];
+	
+	return success;
 }
 
-- (BOOL)accessFileURL:(NSURL *)fileUrl withBlock:(AppSandboxFileAccessBlock)block persistPermission:(BOOL)persist {
+- (BOOL)requestAccessPermissionsForFilePath:(NSString *)filePath persistPermission:(BOOL)persist withBlock:(AppSandboxFileSecurityScopeBlock)block {
+	NSParameterAssert(filePath);
+	
+	NSURL *fileURL = [NSURL fileURLWithPath:filePath];
+	return [self requestAccessPermissionsForFileURL:fileURL persistPermission:persist withBlock:block];
+}
 
-	NSURL *allowedUrl = nil;
+- (BOOL)requestAccessPermissionsForFileURL:(NSURL *)fileURL persistPermission:(BOOL)persist withBlock:(AppSandboxFileSecurityScopeBlock)block {
+	NSParameterAssert(fileURL);
+	
+	[self persistPermissionURL:fileURL];
```

Comment:
@regexident what's the motivation for introducing another persist call here? We're persisting on line 207 of this function, which needs to happen only after permission has been granted which happens in between this line and 207.